### PR TITLE
Support schematic data

### PR DESCRIPTION
**Pull request by ikbalkaya**
Please note - work in progress

This PR adds support for connect records with schemas by Jsonifying Structs and passing it as Ably message as byte[]

### CODE REVIEW

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+public class StructUtils {
+    public static String toJsonString(final Struct struct, Gson gson) {
+        return gson.toJson(structMap(struct));
+    }
+
+    private static Map<String, Object> structMap(final Struct struct) {
+        final Map<String, Object> dataMap = new LinkedHashMap<>(struct.schema().fields().size());
+
+        for (Field field : struct.schema().fields()) {
```
Comment (QuintinWillison) [inline src/main/java/com/ably/kafka/connect/utils/StructUtils.java:20]: `field` can be `final` here.

```suggestion
        for (final Field field : struct.schema().fields()) {
```

**Review by QuintinWillison — COMMENTED** [review]
I've finished a first-pass review of this Draft PR, as requested by @ikbalkaya.

**General Observation, not related strictly to the scope of this PR**: I started, then stopped, commenting on missing `final` modifiers - we could definitely benefit from checkstyle in this codebase to eliminate the need for such [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) (e.g. [FinalParametersCheck](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheck.html)). I've created #74 for this, as we'll solve this another day. 😁 

Generally, in particular when looking at the tests, I found myself somewhat overwhelmed with questions which were not answered by the code in front of me. All classes and methods in Java, including tests, can have JavaDoc commentary attached and I would strongly encourage that you write them. They will always be of use to reviewers and future maintainers alike, but they also have the added benefit of forcing you to reconsider your choices as you write them. In particular in respect of focus, scope, separation of concerns and edge cases (e.g. should it `throw` or not).

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java [hunk 2]
```diff
@@ -35,5 +40,23 @@ public Message getMessage(SinkRecord record) {
...
 
+    private Message messageFromRecord(SinkRecord record) {
+        final String messageName = configValueEvaluator.evaluate(record, sinkConnectorConfig.getString(MESSAGE_CONFIG));
+
+        if (record.valueSchema() == null) {
+            return new Message(messageName, record.value());
+        }
+
+        final Schema valueSchema = record.valueSchema();
+        switch (valueSchema.type()) {
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java:51]: I kept this switch especially to also consider special processing of other schema types. Copied from source, these are all possible types for it 
`INT8,  INT16,INT32, INT64, FLOAT32, FLOAT64,BOOLEAN,STRING, BYTES,ARRAY,MAP,STRUCT`
Any other type than struct - the ably message will directly be built . I did not include these in this PR as other cases than STRUCT were not the case for Avro

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                case ARRAY:
+                    final List<Object> fieldArray = struct.getArray(field.name());
+                    final List<Object> fieldJsonArray = jsonArrayFrom(fieldArray);
+                    dataMap.put(field.name(), fieldJsonArray);
+                    break;
+                case MAP:
+                    final Map<String, Object> jsonMap = mapFrom(struct.getMap(field.name()));
+                    dataMap.put(field.name(), jsonMap);
+                default:
+                    dataMap.put(field.name(), jsonValue(struct.get(field)));
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/utils/StructUtils.java]: It's true, but the idea behind this switch is that any other type except for the special ones described above should be able to resolve into correct json value, including null. I am trying to increase test coverage to include all connect types

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java (hunk 1 above)
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/utils/StructUtils.java]: 00c7e52fed855fdf3d9f266f79eb400978dc820e

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            }
+        }
+        return dataMap;
+    }
+
+
+    private static Map<String, Object> mapFrom(Map<Object, Object> map) {
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/utils/StructUtils.java]: Sorry @QuintinWillison that wasn't very clear. I have made a refactor and wrote some documentation here 4a27b1d5bb179152d820af7555e33f30ccef309e

But basically a method like this is here to extract fields and values from struct schemas and turn them into some map that is in turn can be JSON serializable. Currently 3 special cases are Map, Struct and array - And all other types are passed directly - I am in process of checking whether there is any other special type we need to look at such as when schema type is `BYTES`

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Struct;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class StructUtils {
+    public static String toJsonString(final Struct struct, Gson gson) {
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/utils/StructUtils.java]: 35e44750c8e7a22ebb336a94127a497568227da7

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java]: I have converted switch to an if statement 3d5084998a29708b86ba24f6f01f2a153668a4aa

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/utils/StructUtils.java]: @QuintinWillison Would you have an alternative recommendation for this?

**Comment by ikbalkaya** [comment]
> I've finished a first-pass review of this Draft PR, as requested by @ikbalkaya.
> 
> I started, then stopped, commenting on missing `final` modifiers - we could definitely benefit from checkstyle in this codebase to eliminate the need for such [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) (e.g. [FinalParametersCheck](https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/checks/FinalParametersCheck.html)). I've created #74 for this.
> 
> Generally, in particular when looking at the tests, I found myself somewhat overwhelmed with questions which were not answered by the code in front of me. All classes and methods in Java, including tests, can have JavaDoc commentary attached and I would strongly encourage that you write them. They will always be of use to reviewers and future maintainers alike, but they also have the added benefit of forcing you to reconsider your choices as you write them - i.e. if you can't explain why a thing has been engineered a certain way then it's worth getting yourself into a place so that you can. In particular in respect of focus, scope, separation of concerns and edge cases (e.g. should it `throw` or not).

@QuintinWillison In recent commits I refactor names and documented important classes on what they are doing etc.  I hope they are more clear now.
About tests - I added some explanations to test classes and made test names a bit more clear. I did not want to put too much effort explaining Avro related conversions as these are more to provide some fake schematic data.  I hope it makes more sense now

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java]: Thanks. https://github.com/ably/kafka-connect-ably/commit/3d5084998a29708b86ba24f6f01f2a153668a4aa specifically addresses my question around "`switch` for a single case".

However, what it doesn't address is my second concern, where I question the validity of your code path if the type is anything but `STRUCT` here... e.g. for `BYTES`, or even `MAP`. I would still expect to see an exception to be thrown here in those cases, or am I missing something?

File: src/main/java/com/ably/kafka/connect/utils/StructUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (QuintinWillison) [reply src/main/java/com/ably/kafka/connect/utils/StructUtils.java]: Perhaps more the issue here is the contract for your `jsonValue` method. 🤔 ...

Normally, with any `switch` statement, I will always try to get into a position where the `default` clause is indicative of a failure (i.e. throws an exception). So, here, I would have specific `case` clauses for each and every type you categorically know that you have code paths to support.

Part of the reason for this is that you're beholden to a third party API. As and when new members get added to this set of types then it's likely that your implementation would start failing here, or downstream (i.e. within `jsonValue`'s implementation, or beyond).

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java]: Sorry @QuintinWillison I could have been a bit more clear on explanation

> I did not include these in this PR as other cases than STRUCT were not the case for Avro
Short answer is outside of scope of this PR. 

In case of all other types the message will be passed exactly as they are which was the case before this change. But I think we should address this in subsequent PRs potentially considering Jsonfiying them (TBD)

**Review by QuintinWillison — APPROVED** [review]
Thanks. Looks good to me, though I need to caveat my approval by admitting that I've not had time to get into a position to test this in any kind of a Kafka context (mainly for lack of time, but exacerbated by the fact that I have never used Kafka so would have needed even more time to get myself up-to-speed), therefore my review input has all been based on inspection of your source code changes.

I am hoping @lmars will be able to fill in the gaps in my review-ability by assessing this from his perspective, having had experience with Kafka and having worked on this codebase previously. 🙏

File: README.md [hunk 6]
```diff
@@ -114,6 +114,40 @@ An example cURL command to start the connector in distributed mode is:
     data: {"id":"1543960661:0:2","clientId":"kafka-connect-ably-example","connectionId":"SuJTceISnT","timestamp":1623496744538,"encoding":"base64", "channel":"kafka-connect-ably-example","data":"bWVzc2FnZSAz","name":"sink"}
     ```
 
+#### Publishing messages with schema
+
+Ably Kafka connector supports messages with schema by converting schematic messages to JSON. Produced JSON is then passed as data to Ably message. JSON is produced by eliminating schema fields from the original message.
```
Comment (lmars) [inline README.md:119]: > JSON is produced by eliminating schema fields from the original message.

Can you clarify what you mean by "eliminate"?

File: README.md [hunk 7]
```diff
@@ -114,6 +114,40 @@ An example cURL command to start the connector in distributed mode is:
...
+"key1":{"count":1}
+"key2":{"count":2}
+"key3":{"count":3}
+EOF
+```
+
+You should receive following JSON messages where you subscribed:
+
+```json
+{"count":1}
```
Comment (lmars) [inline README.md:140]: The user will also get the Avro-formatted key base64-encoded in the extras, for example:

```json
{
  "id": "1543960661:0:5",
  "clientId": "kafka-connect-ably-example",
  "connectionId": "sY0XR-p0oW",
  "timestamp": 1653912887495,
  "encoding": "json",
  "extras": {
    "kafka": {
      "key": "AAAAAAEIa2V5Mw=="
    }
  },
  "data": "{\"count\":3}"
}
```

I think it's useful to demonstrate that here.

File: config/docker-compose-worker.properties [hunk 8]
```diff
@@ -8,15 +8,9 @@ bootstrap.servers=kafka:9092
...
-value.converter=org.apache.kafka.connect.converters.ByteArrayConverter
-value.converter.schema.registry.url=http://confluent:8081
-# The internal converter used for offsets and config data is configurable and must be specified,
-# but most users will always want to use the built-in default. Offset and config data is never
-# visible outside of Connect in this format.
-internal.key.converter=org.apache.kafka.connect.json.JsonConverter
-internal.value.converter=org.apache.kafka.connect.json.JsonConverter
-internal.key.converter.schemas.enable=false
-internal.value.converter.schemas.enable=false
+value.converter=io.confluent.connect.avro.AvroConverter
```
Comment (lmars) [inline config/docker-compose-worker.properties:11]: I think `ByteArrayConverter` should remain the default

**Review by lmars — COMMENTED** [review]
I've successfully been able to publish Avro data and get JSON through in Ably, but I found the docs a little hard to follow to accomplish that, so I've made a few documentation related comments.

File: README.md (hunk 6 above)
Comment (ikbalkaya) [reply README.md]: I meant to convey the message to 'get rid of type and other informations from schema' but when I look, I think it's not clear. I made a change here a4942213ba9bd4e516e3ea8afb110f323453a9f4 to hopefully make it a bit more clear but honestly I couldn't find a way to explain this better. I would be open to any suggestions.

File: README.md [hunk 9]
```diff
@@ -114,6 +114,40 @@ An example cURL command to start the connector in distributed mode is:
...
     ```
 
+#### Publishing messages with schema
+
+Ably Kafka connector supports messages with schema by converting schematic messages to JSON. Produced JSON is then passed as data to Ably message. JSON is produced by eliminating schema fields from the original message.
+For example if you use `AvroConverter` and register an Avro schema to your schema registry, you can use the following command providing your own schema and data:
+
+```shell
+ docker-compose exec -T schema-registry kafka-avro-console-producer \
+   --topic topic1 \
```
[the remark this answers is not in the document]
Comment (ikbalkaya) [reply README.md:125]: Added a slightly lengthy description  d334853024aedb3296fbc5395731852f71b1a43b

File: README.md (hunk 7 above)
Comment (ikbalkaya) [reply README.md:141]: Added an example that contains whole message d5edbd101b440c36017a5ebf16ebcc2f95cbe24c

File: config/docker-compose-worker.properties (hunk 8 above)
Comment (ikbalkaya) [reply config/docker-compose-worker.properties]: Good point - Reverted it 7011ca1b67c0537814c4b141f39228514ac310e3

File: README.md [hunk 10]
```diff
@@ -114,6 +114,63 @@ An example cURL command to start the connector in distributed mode is:
     data: {"id":"1543960661:0:2","clientId":"kafka-connect-ably-example","connectionId":"SuJTceISnT","timestamp":1623496744538,"encoding":"base64", "channel":"kafka-connect-ably-example","data":"bWVzc2FnZSAz","name":"sink"}
     ```
 
+#### Publishing messages with schema
+
+Ably Kafka connector supports messages with schema by converting schematic messages to JSON. Produced JSON is then passed as data to Ably message. JSON is produced by mapping schema field names to their values.
```
Comment (lmars) [inline README.md:119]:
```suggestion
The Ably Kafka Connector supports messages which contain schema information by converting them to JSON before publishing them to Ably.
```

File: README.md [hunk 11]
```diff
@@ -114,6 +114,63 @@ An example cURL command to start the connector in distributed mode is:
...
+For example if you use [Kafka Connect Avro Converter](https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter) and register an Avro schema to your [schema registry](https://docs.confluent.io/platform/current/schema-registry/connect.html), you can use the following.
```
Comment (lmars) [inline README.md:120]: Suggested change:

---
For example, if messages on the Kafka topic are serialized using Avro with schemas registered at `https://<your-schema-registry-host>`, then set the following configuration so that those messages are converted from Avro to JSON:

```
value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=https://<your-schema-registry-host>
```

File: README.md [hunk 12]
```diff
@@ -114,6 +114,63 @@ An example cURL command to start the connector in distributed mode is:
...
     ```
 
+#### Publishing messages with schema
+
+Ably Kafka connector supports messages with schema by converting schematic messages to JSON. Produced JSON is then passed as data to Ably message. JSON is produced by mapping schema field names to their values.
+For example if you use [Kafka Connect Avro Converter](https://www.confluent.io/hub/confluentinc/kafka-connect-avro-converter) and register an Avro schema to your [schema registry](https://docs.confluent.io/platform/current/schema-registry/connect.html), you can use the following. 
+
+Based on [docker-compose](/docker-compose.yml),  produce a set of test messages specifying `value.schema` property to a valid [Avro Schema](https://avro.apache.org/docs/current/spec.html) and specifying a valid`schema.registry.url` using the `kafka-avro-console-producer` CLI tool. Note that `kafka-avro-console-producer` tool is a part of schema-registry. Replace `<kafka-topic-name>` with one of the topics set in your configuration file.
```
Comment (lmars) [inline README.md:122]:
```suggestion
If you're running the Ably Kafka Connector locally using Docker Compose as outlined above, then you can use the [`kafka-avro-console-producer` CLI](https://docs.confluent.io/platform/current/tutorials/examples/clients/docs/kafka-commands.html#produce-avro-records) to test producing Avro serialized messages by running the following:
```

File: README.md [hunk 13]
```diff
@@ -114,6 +114,63 @@ An example cURL command to start the connector in distributed mode is:
...
+
+```json
+{"count":2}
+```
+
+```json
+{"count":3}
+```
+
+You will also receive an Avro-formatted key base64 encoded in the extras. For example:
```
Comment (lmars) [inline README.md:152]: I think we should just have one example that shows the data and the keys in extras coming through, rather than one example with just data, and a second with the keys.

File: README.md [hunk 14]
```diff
@@ -114,6 +114,63 @@ An example cURL command to start the connector in distributed mode is:
...
+
+```json
+{
+	"clientId": "Ably-Kafka-Connector",
+	"connectionId": "VSuDXysgaz",
+	"data": {
+		"count": 1
+	},
+	"extras": {
+		"jsonObject": {
```
Comment (lmars) [inline README.md:162]: This `jsonObject` field should not be here.

**Review by lmars — COMMENTED** [review]
I've made some explicit documentation suggestions, but I think it might be worth asking the docs team for their input, since they composed the rest of the README.

File: README.md (hunk 14 above)
Comment (ikbalkaya) [reply README.md:150]: @lmars On subscriber side, this JSOn is produced when I try to JSONify Java Object to JSON and looks like there is a `jsonObject` under extras, which then includes the jSON, From your your comment I assume that this area should be skipped when serializing. Can you please confirm that?
Screenshot 2022-05-31 at 07 37 43

File: README.md (hunk 11 above)
Comment (ikbalkaya) [reply README.md]: 513e2e39a2373faf0be4792da8492b32bd68cd9d

**Comment by ikbalkaya** [comment]
> I've made some explicit documentation suggestions, but I think it might be worth asking the docs team for their input, since they composed the rest of the README.

@lmars Thanks for suggestions on docs, I applied them. There is a question remaining (about 'jsonObject' in Ably message) which needs answering. I added @m-hulbert to review the Readme part of this PR.

File: README.md (hunk 14 above)
Comment (ikbalkaya) [reply README.md]: Removed `jsonObject` node e927f3c0a7163aace284830fc7be75eeabf7b9db but I will wait for the response for above to conclude.

File: README.md [hunk 15]
```diff
@@ -114,6 +114,49 @@ An example cURL command to start the connector in distributed mode is:
     data: {"id":"1543960661:0:2","clientId":"kafka-connect-ably-example","connectionId":"SuJTceISnT","timestamp":1623496744538,"encoding":"base64", "channel":"kafka-connect-ably-example","data":"bWVzc2FnZSAz","name":"sink"}
     ```
 
+#### Publishing messages with schema
+
+The Ably Kafka Connector supports messages which contain schema information by converting them to JSON before publishing them to Ably.
+For example, if messages on the Kafka topic are serialized using Avro with schemas registered at https://<your-schema-registry-host>, then set the following configuration so that those messages are converted from Avro to JSON:
```
Comment (jaley) [inline README.md:120]: While I think this Avro example is useful, we should remember that there's nothing Avro-specific about this connector. We support Kafka Connect schemata, which can be generated using either the Avro, JSON or Protobuf converters. We should probably link to the [Kafka Connect Schema Registry docs](https://docs.confluent.io/platform/current/schema-registry/connect.html) so that users can find the appropriate config for their setup?

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java [hunk 16]
```diff
@@ -35,5 +40,20 @@ public Message getMessage(SinkRecord record) {
...
+        }
 
+        final Schema valueSchema = record.valueSchema();
+        if (valueSchema.type() == Schema.Type.STRUCT) {
+            final String jsonString = StructToJsonConverter.toJsonString((Struct) record.value(), gson);
+            final Message message = new Message(messageName, jsonString);
+            message.encoding = "json";
+            return message;
+        }
+        return new Message(messageName, record.value());
```
Comment (jaley) [inline src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java:57]: If I understand correctly, this will fail if `valueSchema != null` and `valueSchema != STRUCT`, i.e. there is a schema set, but it's not a Struct type, such as Array, String, Integer, etc, as we'll still be passing a Connect data type to `ably-java`, which it won't know how to process? 

Have we made a deliberate decision to only support Struct as the top-level schema? If so, we should probably add a more explicit error, else it'll be raised as a bug should somebody try to send an array or similar. I'm not sure it's really necessary to require this? Are not `1`, `"foo"`, `3.145`, `[1, 2, 3]` etc all valid JSON values? The [docs](https://ably.com/docs/api/realtime-sdk/messages#data) suggest we at least support arrays.

If others agree, I'm happy with us just raising a ticket to add support for other top-level types later rather than delaying this PR.

**Review by jaley — CHANGES_REQUESTED** [review]
The code that's here looks fine, I've made a minor doc improvement comment and suggested either some more explicit error handling for unsupported types, or creating tickets to later support those types (if we think we should) later on. See in-line comments below.

The other thing I wanted to flag was that I'm surprised to see no additions to `DefaultChannelSinkMappingTest` in this PR. That test currently has all the boiler plate ready to set up a `DefaultsinkMapping` instance. We should be able to manually construct instances of [ConnectSchema](https://kafka.apache.org/11/javadoc/org/apache/kafka/connect/data/ConnectSchema.html) and correspecting [Structs](https://kafka.apache.org/11/javadoc/org/apache/kafka/connect/data/Struct.html) to ensure that we can handle incoming data with a schema attached, regardless of whether it's an Avro, Protobuf or JSON schema generating those values originally. It would be good to see cases added to that unit test for things like:

* Given a `SinkRecord` with `Struct` as input, assert that `message.data` contains corresponding JSON
* For unsupported types (currently all non-Struct types at top level, then bytes as scalars), assert that informative errors are thrown
* Other invariants on input data (do we support null?)

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java (hunk 16 above)
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java:57]: @jaley Yes, this is going to only take a Struct as a top level. @QuintinWillison also brought up this question https://github.com/ably/kafka-connect-ably/pull/73#discussion_r881756462 and answered that that was out of the scope of this PR - but looking back I think I might be wrong. My thinking was that there wouldn't be any non-struct top level data with schema, judging from the schemas I tried.  
I now think we need to address this - As we already have our logic that deals with other types in our JSON converter this shouldn't be a big change. WDYT?

Comment (jaley) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java:57]: I think that by far the most common use case will be that the top-level schema is for a Struct, but there's no real need for us to enforce that, as far as I can see. As discussed offline, I think it's fine to just raise a ticket to reconsider this in a separate PR later if we're keen to get support for the majority use case merged ASAP.

This branch, however (line 57), I think will *always* throw an unhelpful exception to the end user? We know that the record has a schema and that schema is not a Struct. The value is therefore a Connect value, which `ably-java` won't know how to handle. Is that correct? If so, and we want to just revisit handling other top-level data types later, I think we should at least change this behaviour so that we throw an exception explicitly describing the problem (unexpected top-level data type, etc), rather than letting `ably-java` throw something more mysterious.

Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java:57]: That makes sense @jaley I will throw an explicit ConnectException as I do in other places.

File: README.md (hunk 14 above)
Comment (lmars) [reply README.md]: > this JSOn is produced when I try to JSONify Java Object to JSON

Ok, I'm not sure what Java is doing there, but `jsonObject` doesn't appear in the underlying message coming out of Ably.

Easiest way to see raw messages is to subscribe via SSE as shown in the other example, and you'll see there won't be any `jsonObject` in there.

File: README.md [hunk 17]
```diff
@@ -114,6 +114,49 @@ An example cURL command to start the connector in distributed mode is:
...
+
+value.converter=io.confluent.connect.avro.AvroConverter
+value.converter.schema.registry.url=https://<your-schema-registry-host>
+`
```
Comment (lmars) [inline README.md:124]: This should be surrounded with three backticks ("```") to make it a code block

**Review by lmars — APPROVED** [review]
I'm happy to approve based on the fact that following the documentation results in me being able to publish Avro-serialized data and get JSON out of Ably, but I'll leave approval of the Java code to other reviewers.

File: src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java (hunk 16 above)
Comment (ikbalkaya) [reply src/main/java/com/ably/kafka/connect/mapping/DefaultMessageSinkMapping.java]: @jaley I made the change here cf1cdfbafa7b26c32c90e01156dbe5405dfb25b8 - however I realized our previous test were broken - because we already support bytes and also strings. I added support for them in subsequent commits

File: README.md (hunk 17 above)
Comment (ikbalkaya) [reply README.md]: 63ac487f62b51c8b32d382202735ab9e4ed48039

Comment (QuintinWillison) [reply README.md]: Thanks. I'm resolving this conversation on @lmars' behalf as I can see that @ikbalkaya has fixed this.
